src/Features/Core/Portable/ConvertForToForEach/AbstractConvertForToForEachCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,503 @@
+            findAndReplaceMatches(forStatement);
+
+            // Finally, remove the declaration statement if we found one.  Move all its leading
+            // trivia to the next statement.
+            if (declarationStatement != null)
+            {
+                editor.RemoveNode(declarationStatement,
+                    SyntaxGenerator.DefaultRemoveOptions | SyntaxRemoveOptions.KeepLeadingTrivia);
+            }
+
+            var options = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
+            editor.ReplaceNode(
+                forStatement,
+                (currentFor, _) => this.ConvertForNode(
+                    (TForStatementSyntax)currentFor, typeNode, foreachIdentifier,
+                    collectionExpression, iterationType, options));
+
+            return document.WithSyntaxRoot(editor.GetChangedRoot());
+
+            // local functions
+            (TTypeNode, SyntaxToken, TStatementSyntax) tryDeconstructInitialDeclaration()
+            {
+                if (bodyStatements.Count >= 1)
+                {
+                    var firstStatement = bodyStatements[0];
+                    if (syntaxFacts.IsLocalDeclarationStatement(firstStatement))
+                    {
+                        var variables = syntaxFacts.GetVariablesOfLocalDeclarationStatement(firstStatement);
+                        if (variables.Count == 1)
+                        {
+                            var firstVariable = (TVariableDeclaratorSyntax)variables[0];
+                            if (IsValidVariableDeclarator(firstVariable))
+                            {
+                                var firstVariableInitializer = syntaxFacts.GetValueOfEqualsValueClause(
+                                    syntaxFacts.GetInitializerOfVariableDeclarator(firstVariable));
+                                if (syntaxFacts.AreEquivalent(firstVariableInitializer, indexExpression))
+                                {
+                                    var type = (TTypeNode)syntaxFacts.GetTypeOfVariableDeclarator(firstVariable)?.WithoutLeadingTrivia();
+                                    var identifier = syntaxFacts.GetIdentifierOfVariableDeclarator(firstVariable);
+                                    var statement = firstStatement;
+                                    return (type, identifier, statement);
+                                }
+                            }
+                        }
+                    }
+                }
+
+                return default;
+            }
+
+            void findAndReplaceMatches(SyntaxNode current)
+            {
+                if (syntaxFacts.AreEquivalent(current, indexExpression))
+                {
+                    // Found a match.  replace with iteration variable.
+                    var replacementToken = foreachIdentifierReference;
+
+                    if (semanticFacts.IsWrittenTo(semanticModel, current, cancellationToken))
```

Comment:
why not use data flow API to check written and captured variable?

Reply:
Because it's not the variable that's written to.  It's the *expression*.  i.e you had:

```this.x.y.z[i] = 10```

You're looking to see how ```this.x.y.z[i]``` is used.  Not how some variable is used.

Reply:
you could create semanicModel.GetDataFlowAnalysis() from foreach body. and then pass that down. and check dataflow.WrittenTo.Contains or dataFlow.Captured.Contains

Reply:
DataFlow is *symbol* based.  It is not node/expression based.  I cannot ask "was ```this.x.y.z[i]``` written to in the expression ```this.x.y.z[i]++```".

Reply:
I see. @mavasani is it something new data flow api can support?

Reply:
talked with @mavasani looks like one he is building based on CFG @AlekseyTs is creating can identify this as well!

Reply:
@mavasani Can you please include me in the loop on the dataflow API?  Note: otehr IDE features (like FindReferences) wants to know about this information as well so it can properly categorize references.

Reply:
> Can you please include me in the loop on the dataflow API? Note: otehr IDE features (like FindReferences) wants to know about this information as well so it can properly categorize references.

@CyrusNajmabadi Sure. Current work in the compilers is primarily towards ControlFlowGraph API. Subsequently, we will move to dataflow APIs.